### ISSUE TITLE
torch.nn.DataParallel causes incorrect gradients

### ISSUE THREAD

**Issue by xiaoguai0992** [pytorch/pytorch#23938]
# Bug Report
## Issue description

I have a model which has two nn.Conv2d modules, and I only use the first one of them in 'forward'.

In general, after executing 'loss.backward()', all weights' gradients of the second Conv2d (the unused one) should be 'None'.

Without nn.DataParallel, I got the correct result (conv2.weight.grad is None).

However with nn.DataParallel, the conv2.weight.grad is a zero tensor instead of None. So that if I run optimizer.step() after backward, weight_decay and momentum will be accumulated for the unused parameters, which causes unexpected results. I hope those gradients of unused parameters keep 'None' instead of a zero tensor. 

I have a sample fix for this issue temporarily, but that might cause other problems (when real p.grad = zeros).
```python
loss.backward()
for p in model.parameters():
    if torch.sum(torch.abs(p)) == 0.0:
        p.grad = None
optimizer.step()
```

So why does this problem occur? And how to fix it correctly?

## Code example

See https://gist.github.com/xiaoguai0992/db8742c3fa7a5e02be36e64180693752

The output of the code should be:
```bash
Testing non-dataparallel.
conv1.weight, p.grad is None = False
conv1.bias, p.grad is None = False
conv2.weight, p.grad is None = True
conv2.bias, p.grad is None = True
Testing dataparallel
module.conv1.weight, p.grad is None = False
module.conv1.bias, p.grad is None = False
module.conv2.weight, p.grad is None = False
module.conv2.bias, p.grad is None = False
Testing repaired version
module.conv1.weight, p.grad is None = False
module.conv1.bias, p.grad is None = False
module.conv2.weight, p.grad is None = True
module.conv2.bias, p.grad is None = True
```

## System Info
PyTorch version: 1.1.0
Is debug build: No
CUDA used to build PyTorch: 9.0.176

OS: Ubuntu 16.04.6 LTS
GCC version: (Ubuntu 5.4.0-6ubuntu1~16.04.11) 5.4.0 20160609
CMake version: version 3.5.1

Python version: 3.5
Is CUDA available: Yes
CUDA runtime version: Could not collect
GPU models and configuration: 
GPU 0: GeForce RTX 2080 Ti
GPU 1: GeForce RTX 2080 Ti
GPU 2: GeForce RTX 2080 Ti
GPU 3: GeForce RTX 2080 Ti

Nvidia driver version: 418.40.04
cuDNN version: Could not collect

Versions of relevant libraries:
[pip3] numpy==1.16.2
[pip3] torch==1.1.0
[pip3] torchvision==0.2.2.post3
[conda] Could not collect

**Comment by mrshenli** [comment]
Hi @xiaoguai0992, it might be caused by the same issue as #14960, which is fixed in #22983. Can you try if using the [nightly](https://anaconda.org/pytorch/pytorch-nightly) build fixes the problem?

**Comment by xiaoguai0992** [comment]
> Can you try if using the [nightly](https://anaconda.org/pytorch/pytorch-nightly) build fixes the problem?

I found that the nightly version of pytorch is 1.0.0.dev20190328 in your reply.(However the newest version is 1.1.0) Is that nightly version correct? If correct, then nightly version cannot solve my problem.

**Comment by mrshenli** [comment]
@xiaoguai0992 Sorry about the delay. 1.0.0.dev20190328 does not contain the fix. Can you try v1.2?

```
conda install pytorch=1.2.0 cuda90 -c pytorch
```

**Comment by xiaoguai0992** [comment]
> 
> ```
> conda install pytorch=1.2.0 cuda90 -c pytorch
> ```

Hi, it seems that pytorch 1.2.0 is not compatible with cuda9.0，so that I cannot test this issue under cuda9.0.  But cuda 10 is ok. 

However, updating pytorch to 1.2.0, this problem still exists.

**Comment by Wu-Hai** [comment]
> > ```
> > conda install pytorch=1.2.0 cuda90 -c pytorch
> > ```
> 
> 
> Hi, it seems that pytorch 1.2.0 is not compatible with cuda9.0，so that I cannot test this issue under cuda9.0.  But cuda 10 is ok.
> 
> 
> However, after updating pytorch to 1.2.0, this problem still exists.

Hi, great thanks for your report this issue. I meet the same problem when using DataParallel to calculate the backward gradient. The grad None will be replaced by 0 (some thing like this).  I wonder if this problem has been solved in the latest pytorch version, aka(1.9.0)
